composer.json

```diff
@@ -19,7 +19,8 @@
         "ext-tokenizer": "*",
         "composer/semver": "^3.2",
         "composer/xdebug-handler": "^3.0.3",
-        "doctrine/annotations": "^1.13",
+        "doctrine/annotations": "^1.13 || ^2",
+        "doctrine/lexer": "^1 || ^2",
```

Comment:
why we need to explicitly mention this peer dependency ?

Reply:
PHP CS Fixer does not rely on `Doctrine\Common\Lexer` directly. why we need to mention it [or why we need to prevent v3]?
shouldn't it be covered by whatever using this library?

If behaviour of Doctrine/Annotation v2 is different, depends on version of it's own dependency [lexer], sounds like issue on Annotation package, imho.

Reply:
> PHP CS Fixer does not rely on `Doctrine\Common\Lexer` directly.

It kinda does by using `Doctrine\Common\Annotations\DocLexer` which extends a class from the Lexer library.

> why we need to mention it

Because Annotations 2 would allow the installation of Lexer 3 which requires some code changes on CS Fixer's side. I can make those adjustments if you want.

The main issue is that Lexer 2 replaces the associative array structure of tokens with a dedicated `Token` class. We access that array structure here for instance:

https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/fb41eb66a32dc186f8068c134055701ac272b217/src/Doctrine/Annotation/Tokens.php#L80

In Lexer 1, this property needs to be accessed via `$token['type']` while Lexer 3 requires us to write this as `$token->type` instead. Lexer 2 allows both ways by implementing `ArrayAccess` as a compat layer. There are no new features in Lexer 3 and Doctrine has not even transitioned their own projects to Lexer 3 yet, so there's no urgency in supporting it just yet.

Our options are:

1. Keep this PR as is and work on Lexer 3 compat with a follow-up. Annotations 1 blocks the installation of Lexer 3 anyway, so this new line in composer.json wouldn't generate a conflict that isn't already there.
2. Remove the new constraint and support Lexer 1, 2 and 3: This will make the codebase a little more complicated, but I think I can bury that complexity in a private method.
3. Require Lexer `^2 || ^3`. No multi-version hacks and the codebase is future-proof, but might cause downstream conflicts at the moment.

Reply:
I am totally shocked that I can install annotations v2 and get totally different interface as return [sometimes assoc array, sometimes DTO] ;/

Reply:
https://packagist.org/packages/doctrine/annotations/stats#major/all
v2 not used at all, yet

https://packagist.org/packages/doctrine/lexer/stats#major/all
v1, a bit v2, no v3 yet

IMHO:
- annotations: v1 || v2, lexer: v1 || v2
- let's prepare 2 PRs. one that is allowing lexer v1 || v2 (this PR), other that is allowing only v2.
  - let's release first PR as vX.Y.Z, and 2nd PR as vX.Y.(Z+1). then, we see how many of our users will stuck with Z and not able to upgrade to Z+1.
    - and in case of need, PR with v2 only will be easy to revert, if there will be huge gap between Z and Z+1
    - and if all good, we can extend it to v2 || v3 and use future-ready interface only

Reply:
Sounds like a plan. I'll prepare the second PR.

Reply:
I've created #6729 and #6730 as follow-ups.

Reply:
Thanks !
It will take me a while to merge [ I want to merge first the bugfixes first, and then release those deps compatibility on top ]

Reply:
I see. I'm currently submitting this kind of PR to several open-source libs to push the adoption of the new Lexer and Annotations releases. Some of them are currently impeded by CS-Fixer.

Merging (and releasing) this PR at least would enable me to continue my work.

Reply:
> I am totally shocked that I can install annotations v2 and get totally different interface as return [sometimes assoc array, sometimes DTO] ;/


That's not the case actually, with annotations v2 you always get a `Token` object. But you're right in that it's a different interface, because that object implements `ArrayAccess` with lexer v2, and no longer does with lexer v3.